db/migrate/20150525081931_remove_duplicate_tokens.rb

```diff
@@ -0,0 +1,28 @@
+class RemoveDuplicateTokens < ActiveRecord::Migration
+  def up
+    # only managed hosts should have tokens
+    Token.where(:id => Token.joins("left outer join hosts on hosts.id = tokens.host_id").
+                             where("hosts.type != 'Host::Managed'").pluck('tokens.id')).delete_all
+
+    # deleting duplicate tokens
+    hosts_with_duplicate_tokens = Token.having('count(*) > 1').group(:host_id).select('host_id')
+    existing_tokens = Host.where(:id => hosts_with_duplicate_tokens).map(&:token).compact
+    if existing_tokens.empty?
+      Token.delete_all
+    else
+      Token.where('id not in (?)', existing_tokens).delete_all
```

Comment:
Perhaps say "where id not in (existing_tokens) and host_id in (hosts_with_duplicate_tokens)"?